werkzeug/routing.py

```diff
@@ -794,38 +801,307 @@ def match(self, path, method=None):
+                ops += rv
+                peak_stack = max(stack + ps, peak_stack)
+                stack += 1
+                if append_unknown:
+                    ps, rv = self.emit_build(len(ops),
+                                             url_ops,
+                                             append_unknown,
+                                             encode_query_vars,
+                                             argcount)
+                else:
+                    ps, rv = self.emit_build(len(ops), url_ops)
+                ops += rv
+                peak_stack = max(stack + ps, peak_stack)
+                ops += self.build_op('BUILD_TUPLE', 2)
+            ops += self.build_op('RETURN_VALUE')
+            code_args = [argcount,
+                         len(self.var),
+                         peak_stack + len(self.var),
+                         flags,
+                         ops,
+                         tuple(self.consts),
+                         (),
+                         tuple(self.var),
+                         'generated',
+                         '<builder:%r>' % self.rule.rule,
+                         1,
+                         b'']
+            if sys.version_info >= (3,):
+                code_args[1:1] = [0]
+            else:
+                code_args[4] = str(code_args[4])
+            co = types.CodeType(*code_args)
+            fn = types.FunctionType(co, {}, None, self.argdefs)
+            return fn
+
+    def compile_builder(self, append_unknown=True):
+        return self.BuilderCompiler(self).compile(append_unknown)
+
     def build(self, values, append_unknown=True):
         """Assembles the relative url for that rule and the subdomain.
         If building doesn't work for some reasons `None` is returned.
 
         :internal:
         """
-        tmp = []
-        add = tmp.append
-        processed = set(self.arguments)
-        for is_dynamic, data in self._trace:
-            if is_dynamic:
-                try:
-                    add(self._converters[data].to_url(values[data]))
-                except ValidationError:
-                    return
-                processed.add(data)
+        try:
+            if append_unknown:
+                return self._build_u(**values)
             else:
-                add(url_quote(to_bytes(data, self.map.charset), safe='/:|+'))
-        domain_part, url = (u''.join(tmp)).split(u'|', 1)
-
-        if append_unknown:
-            query_vars = MultiDict(values)
-            for key in processed:
-                if key in query_vars:
-                    del query_vars[key]
-
-            if query_vars:
-                url += u'?' + url_encode(query_vars, charset=self.map.charset,
-                                         sort=self.map.sort_parameters,
-                                         key=self.map.sort_key)
-
-        return domain_part, url
+                return self._build(**values)
+        except TypeError:
```

Comment:
I'm assuming this replaces the `except ValidationError` from the previous code, but why has it changed to `TypeError`? `BuilderCompiler` doesn't appear to hanlde `ValidationError` either.

Reply:
Functions raise `TypeError` when they're called with improper arguments, so I caught it—with hindsight, I think I'd missed that `Rule.build` is internal and doesn't need to handle such cases, so it could go. As for `ValidationError`, converters never raise it when going from Python to URL anyway.

Reply:
The `NumberConverter` raises a `ValidationError` in both `to_python` and `to_url`. The fact that it was being handled suggests it was part of the API even if few of the built-in converters used it.

Reply:
I'm not seeing where, but I'm sure putting it back can't hurt. Will do.

Reply:
Well, you're right, I was mistaken about `NumberConverter.to_python`, but I think we should still handle the case.

Reply:
Sure, done 7d10b585a5b499f82204313ce426c983a155e2a7